core/tests/performance_framework/selenium_driver.py

```diff
@@ -0,0 +1,237 @@
+    """Fetches performance data for locally served Oppia pages using Selenium
+    and Browsermob-proxy.
+
+    Selenium is used to programitically interact with a browser.
+    Browsermob-proxy is used to capture HTTP Archive (referred to as HAR) data.
+
+    The HTTP Archive format is a JSON-formatted archive file format used for
+    logging a web browser's interaction with a site. It contains detailed
+    performance data, including information about page loading and displaying
+    and per-resource statistics. Each entry contains the URL requested and
+    request and response headers. For additional details, please see:
+    https://dvcs.w3.org/hg/webperf/raw-file/tip/specs/HAR/Overview.html
+    """
+
+    # Chrome is the default browser that selenium will use for extracting
+    # performance stats.
+    DEFAULT_BROWSER_SOURCE = 'chrome'
+    SUPPORTED_BROWSER_SOURCES = ['chrome', 'firefox']
+
+    def __init__(self, browser=DEFAULT_BROWSER_SOURCE):
+        if browser in self.SUPPORTED_BROWSER_SOURCES:
+            self.browser = browser
+        else:
+            error_msg = 'Unsupported browser specified: %s' % browser
+            raise ValueError(error_msg)
+
+        self.server = self.driver = self.proxy = None
+
+    def get_har_dict(self, page_url):
+        """Retrieves an HTTP Archive (HAR) dict for a given page URL.
+
+        The HAR dict is retrieved using the proxy server by recording the
+        communication between the client and server.
+        """
+        with Xvfb() as _xvfb:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
+
+            self.proxy.new_har(page_url, options={'captureHeaders': True})
+
+            self.driver.get(page_url)
+
+            # Wait for the complete page to load, otherwise XHR requests made
+            # after page load will not be recorded.
+            time.sleep(WAIT_DURATION)
+
+        har_dict = self.proxy.har
+
+        self._stop_proxy()
+        self._stop_driver()
+
+        return har_dict
+
+    def get_har_dict_cached_state(self, page_url):
+        """Retrieve HTTP Archive (HAR) or page session stats for a page URL
+        while simulating a cached state i.e, a return user.
+
+        To get it we require the use of a proxy server as we need to record the
+        communication between server and client.
+        """
+        with Xvfb() as _xvfb:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
+
+            # Fetch the page once. This leads to caching of various resources.
+            self.driver.get(page_url)
+
+            # Start recording har data for the next page fetch.
+            self.proxy.new_har(page_url, options={'captureHeaders': True})
+            self.driver.get(page_url)
+
+            # Wait for the complete page to load, otherwise XHR requests made
+            # after page load will not be recorded.
+            time.sleep(WAIT_DURATION)
```

Comment:
Unfortunately, XHR requests made after the page load (eg. library page makes a call to /libraryindexhandler) weren't getting logged correctly. Even though the request was made before "onload", but the "onload" event was fired before it finished.